qiskit/qasm3/ast.py

```diff
@@ -697,17 +698,26 @@ class ForLoopStatement(Statement):
             | "[" Range "]"
     """
 
-    __slots__ = ("indexset", "parameter", "body")
+    __slots__ = ("indexset", "parameter", "body", "type")
 
     def __init__(
         self,
         indexset: Union[Identifier, IndexSet, Range],
         parameter: Identifier,
         body: ProgramBlock,
+        type_: Optional[ClassicalType] = None,
     ):
         self.indexset = indexset
         self.parameter = parameter
         self.body = body
+        self.type = type_ if type_ is not None else IntType()
```

Comment:
The AST constructor shouldn't really be making assumptions about types.  It's either got a type, or it hasn't.  If you very much want to try and provide a deprecation warning, then we'd just maintain the `None` here, and have the printer not print a type if it's `None`, rather than defaulting to something arbitrary.

Really though, it's fine just to immediately require that this is a non-optional - the class isn't public and shouldn't be constructed by anything other than Qiskit internals.

Reply:
But not printing a type in OpenQASM3 is invalid now isn't it?